File: nhglib-core/src/io/github/voidzombie/nhglib/runtime/fsm/states/assets/AssetStateLoading.java
package io.github.voidzombie.nhglib.runtime.fsm.states.assets;

import com.badlogic.gdx.ai.fsm.State;
import com.badlogic.gdx.ai.msg.Telegram;
import com.badlogic.gdx.utils.Array;
import io.github.voidzombie.nhglib.NHG;
import io.github.voidzombie.nhglib.assets.Asset;
import io.github.voidzombie.nhglib.assets.Assets;
import io.github.voidzombie.nhglib.runtime.fsm.base.AssetsStates;
import io.reactivex.Observable;
import io.reactivex.functions.Action;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Predicate;

/**
 * Created by Fausto Napoli on 08/12/2016.
 */
public class AssetStateLoading implements State<Assets> {
    @Override
    public void enter(Assets assets) {
        NHG.logger.log(this, "Asset manager is loading.");
    }

    @Override
    public void update(Assets assets) {
        if (assets.assetManager.update()) {
            assets.fsm.changeState(AssetsStates.IDLE);

            assets.assetLoadingFinished();
            publishLoadedAssets(assets);
        }
    }

    @Override
    public void exit(Assets assets) {
        NHG.logger.log(this, "Asset manager has finished loading.");
    }

    @Override
    public boolean onMessage(Assets entity, Telegram telegram) {
        return false;
    }

    private void publishLoadedAssets(final Assets assets) {
        Array<Asset> assetsCopy = new Array<>(assets.getAssetList());

        Observable.fromIterable(assetsCopy)
                .filter(new Predicate<Asset>() {
                    @Override
                    public boolean test(Asset asset) throws Exception {
                        return assets.assetManager.isLoaded(asset.source);
                    }
                })
                .subscribe(new Consumer<Asset>() {
                    @Override
                    public void accept(Asset asset) throws Exception {
                        NHG.logger.log(
                                this,
                                NHG.strings.messages.assetLoaded,
                                asset.source);

                        assets.assetLoaded(asset);
                        assets.dequeueAsset(asset);
                    }
                });
    }
}

File: nhglib-core/src/io/github/movementspeed/nhglib/core/fsm/states/assets/AssetStateLoading.kt
package io.github.movementspeed.nhglib.core.fsm.states.assets

import com.badlogic.gdx.ai.fsm.State
import com.badlogic.gdx.ai.msg.MessageManager
import com.badlogic.gdx.ai.msg.Telegram
import com.badlogic.gdx.utils.Array
import io.github.movementspeed.nhglib.assets.Asset
import io.github.movementspeed.nhglib.assets.Assets
import io.github.movementspeed.nhglib.core.fsm.base.AssetsStates
import io.github.movementspeed.nhglib.utils.data.Strings
import io.github.movementspeed.nhglib.utils.debug.NhgLogger
import io.reactivex.Observable
import io.reactivex.functions.Consumer
import io.reactivex.functions.Predicate

/**
 * Created by Fausto Napoli on 08/12/2016.
 */
class AssetStateLoading : State<Assets> {
    override fun enter(assets: Assets) {
        NhgLogger.log(this, "Asset manager is loading.")
    }

    override fun update(assets: Assets) {
        if (assets.updateAssetManagers()) {
            assets.fsm.changeState(AssetsStates.IDLE)
            MessageManager.getInstance().dispatchMessage(AssetsStates.ASSETS_GC)

            assets.assetLoadingFinished()
            publishLoadedAssets(assets)
        }
    }

    override fun exit(assets: Assets) {
        NhgLogger.log(this, "Asset manager has finished loading.")
    }

    override fun onMessage(entity: Assets, telegram: Telegram): Boolean {
        return false
    }

    private fun publishLoadedAssets(assets: Assets) {
        val assetsCopy = Array(assets.assetQueue!!)

        Observable.fromIterable(assetsCopy)
                .filter { asset -> assets.isAssetLoaded(asset) }
                .subscribe(object : Consumer<Asset> {
                    override fun accept(asset: Asset) {
                        NhgLogger.log(
                                this,
                                Strings.Messages.assetLoaded,
                                asset.source)

                        assets.assetLoaded(asset)
                        assets.dequeueAsset(asset)
                    }
                })
    }
}
